### PR TITLE
Cancel long poll requests when failover 

### PR DESCRIPTION
**Pull request by yux0**
## What changed?
Stop task queue manager to reject long pull requests

## Why?
Scenario:
1. worker connect to the source cluster and poll new task (no task in the task queue).
2. namespace handover
3. no poller polls task from the target cluster until the connections from step 1 times out (max: 1 minutes).
Before the connection re-establishes, users may see task schedule delay.

### CODE REVIEW

File: service/matching/physical_task_queue_manager.go [hunk 1]
```diff
@@ -324,6 +324,17 @@ func (c *physicalTaskQueueManagerImpl) Start() {
...
 	c.partitionMgr.engine.updatePhysicalTaskQueueGauge(c, 1)
+	c.namespaceRegistry.RegisterStateChangeCallback(c.queue, func(ns *namespace.Namespace, deletedFromDb bool) {
+		if deletedFromDb {
+			return
+		}
+
+		if ns.IsGlobalNamespace() &&
+			ns.ReplicationPolicy() == namespace.ReplicationPolicyMultiCluster &&
+			ns.ActiveClusterName() != c.clusterMeta.GetCurrentClusterName() {
+			c.UnloadFromPartitionManager(unloadCauseNamespaceFailover)
```
Comment (yux0) [inline service/matching/physical_task_queue_manager.go:335]: @dnr I am not 100% if I should stop the task queue partition manager.

Comment (dnr) [reply service/matching/physical_task_queue_manager.go:335]: namespace state change callbacks are global, this isn't specific to a particular namespace. i.e. you need one of these at the matching engine level and it should do something like this https://github.com/temporalio/temporal/blob/main/service/matching/matching_engine.go#L281
(except you might not need the double-check and delay)

File: common/collection/sync_map.go [hunk 2]
```diff
@@ -47,6 +47,24 @@ func (m *SyncMap[K, V]) Get(key K) (value V, ok bool) {
 	return
 }
 
+func (m *SyncMap[K, V]) GetOrSet(key K, value V) (v V, exist bool) {
```
Comment (dnr) [inline common/collection/sync_map.go:50]: not part of this PR, but can you change the `*sync.RWMutex` to `sync.RWMutex` (no pointer)?

File: common/collection/sync_map.go [hunk 3]
```diff
@@ -68,3 +86,13 @@ func (m *SyncMap[K, V]) Pop(key K) (value V, ok bool) {
 	}
 	return value, ok
 }
+
+func (m *SyncMap[K, V]) KeySet() []K {
+	m.Lock()
+	defer m.Unlock()
```
Comment (dnr) [inline common/collection/sync_map.go:92]: can be RLock?

File: common/collection/sync_map.go [hunk 4]
```diff
@@ -68,3 +86,13 @@ func (m *SyncMap[K, V]) Pop(key K) (value V, ok bool) {
...
+	keys := make([]K, 0, len(m.contents))
+	for k := range m.contents {
+		keys = append(keys, k)
+	}
+	return keys
```
Comment (dnr) [inline common/collection/sync_map.go:97]:
```suggestion
	return maps.Keys(m.contents)
```
?

(using `"golang.org/x/exp/maps"`, not the stdlib)

File: common/collection/sync_map_test.go
Comment (dnr) [inline common/collection/sync_map_test.go:1]: add simple test for new functions?

File: service/frontend/workflow_handler.go [hunk 5]
```diff
@@ -235,6 +240,26 @@ func (wh *WorkflowHandler) Start() {
...
+			if deletedFromDb {
+				return
+			}
+
+			if ns.IsGlobalNamespace() &&
+				ns.ReplicationPolicy() == namespace.ReplicationPolicyMultiCluster &&
+				ns.ActiveClusterName() != wh.clusterMetadata.GetCurrentClusterName() {
+				pollers, ok := wh.outstandingPollers.Get(ns.ID().String())
+				if ok {
+					for _, key := range pollers.KeySet() {
```
Comment (dnr) [inline service/frontend/workflow_handler.go:254]: hmm.. since you need to get both keys and values, maybe the SyncMap interface you want is actually `GetAndClear`, that returns the current map and then resets it to an empty one? then you can just through and clear them all without additional locking.

File: service/frontend/workflow_handler.go [hunk 6]
```diff
@@ -4743,6 +4780,31 @@ func (wh *WorkflowHandler) cancelOutstandingPoll(ctx context.Context, namespaceI
 	return true
 }
 
+func (wh *WorkflowHandler) registerOutstandingPollContext(
+	ctx context.Context,
+	pollerID string,
+	namespaceID string,
+) context.Context {
+
+	if pollerID != "" {
```
Comment (dnr) [inline service/frontend/workflow_handler.go:4789]: do you need this check? all the callers clearly create a non-empty pollerID?

File: service/frontend/workflow_handler.go [hunk 7]
```diff
@@ -4743,6 +4780,31 @@ func (wh *WorkflowHandler) cancelOutstandingPoll(ctx context.Context, namespaceI
...
+		childCtx, cancel := context.WithCancel(ctx)
+		nsPollers, _ := wh.outstandingPollers.GetOrSet(namespaceID, collection.NewSyncMap[string, context.CancelFunc]())
```
Comment (dnr) [inline service/frontend/workflow_handler.go:4791]: hmm, I don't like the idea that it's creating a map and throwing it out almost all the time. maybe it should pass in a factory function instead?

File: service/frontend/workflow_handler.go [hunk 8]
```diff
@@ -4743,6 +4780,31 @@ func (wh *WorkflowHandler) cancelOutstandingPoll(ctx context.Context, namespaceI
...
+	return ctx
+}
+
+func (wh *WorkflowHandler) unregisterOutstandingPollContext(
+	pollerID string,
+	namespaceID string,
+) {
+	nsPollers, ok := wh.outstandingPollers.Get(namespaceID)
+	if ok {
+		nsPollers.Delete(pollerID)
```
Comment (dnr) [inline service/frontend/workflow_handler.go:4804]: shouldn't you always call `cancel()` after you create a context with `WithCancel`? this causes a leak otherwise?

File: common/collection/sync_map.go (hunk 2 above)
Comment (yux0) [reply common/collection/sync_map.go:50]: updated

File: service/frontend/workflow_handler.go (hunk 8 above)
Comment (yux0) [reply service/frontend/workflow_handler.go:4804]: yes. called cancel()

File: service/frontend/workflow_handler.go (hunk 5 above)
Comment (yux0) [reply service/frontend/workflow_handler.go:254]: change to popAll

File: service/frontend/workflow_handler.go (hunk 7 above)
Comment (yux0) [reply service/frontend/workflow_handler.go:4791]: changed to call get first then getorset.

File: service/matching/physical_task_queue_manager.go (hunk 1 above)
Comment (yux0) [reply service/matching/physical_task_queue_manager.go]: not going to use this approach.

File: common/collection/sync_map.go [hunk 9]
```diff
@@ -24,18 +24,27 @@
...
 
+type KeyValuePair[K, V any] struct {
+	Key   K
+	Value V
+}
+
 func NewSyncMap[K comparable, V any]() SyncMap[K, V] {
 	return SyncMap[K, V]{
-		RWMutex:  &sync.RWMutex{},
+		RWMutex:  sync.RWMutex{},
```
Comment (dnr) [inline common/collection/sync_map.go:45]: don't need to initialize this
```suggestion
```

File: common/collection/sync_map.go [hunk 10]
```diff
@@ -68,3 +95,17 @@ func (m *SyncMap[K, V]) Pop(key K) (value V, ok bool) {
...
+	var result []KeyValuePair[K, V]
+	for k, v := range m.contents {
+		result = append(result, KeyValuePair[K, V]{
+			Key:   k,
+			Value: v,
+		})
+	}
+	maps.Clear(m.contents)
+	return result
+}
```
Comment (dnr) [inline common/collection/sync_map.go]: a simpler/faster approach would be:
```suggestion
func (m *SyncMap[K, V]) PopAll() map[K]V {
	m.Lock()
	defer m.Unlock()
	contents := m.contents
	m.contents = make(map[K]V)
	return contents
}
```
this way is okay too if you want

Comment (dnr) [reply common/collection/sync_map.go]: if you do use the slice approach, it's faster to do `clear(m.contents)` (new in go 1.21) than to delete each key individually

File: common/collection/sync_map.go (hunk 9 above)
Comment (yux0) [reply common/collection/sync_map.go]: removed

**Review by dnr — APPROVED** [review]

File: common/collection/sync_map.go (hunk 10 above)
Comment (yux0) [reply common/collection/sync_map.go]: Thank you. Didn't know it was there.
